### PR TITLE
Adding HNSW-QG AVX512f version

### PR DESCRIPTION
**Pull request by Patrick-H-Chen**
*Issue #, if available:*

Description of changes:

This change reflects the implementation of product quantization with AVX512f SIMD on top of HNSW to perform fast quantization-based graph search. Following are two major changes:

1) Adding a ProductQuantizer with 4 bits local centroids.
2) A HNSWProductQuantizer4Bits struct which leverages ProductQuantization as the distance calculation instead of original full L2 distance.
 

 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pecos/core/ann/hnsw.hpp [hunk 1]
```diff
@@ -238,6 +243,142 @@ namespace ann {
         }
     };
 
+    template<class FeatVec_T>
```
Comment (OctoberChang) [inline pecos/core/ann/hnsw.hpp:246]: Do we still need the template syntax, if we are restricting to dense vector only?

File: pecos/core/ann/hnsw.hpp [hunk 2]
```diff
@@ -870,5 +1009,343 @@ namespace ann {
...
+                throw std::invalid_argument("Inconsistent HNSW_T: hnsw_t_cur = " + hnsw_t_cur  + " hnsw_t_cur = " + hnsw_t_inp);
+            }
+            return j_param;
+        }
+
+        void save_config(const std::string& filepath) const {
+            nlohmann::json j_params = {
+                {"hnsw_t", pecos::type_util::full_name<HNSWProductQuantizer4Bits>()},
+                {"version", "v1.0"},
+                {"train_params", {
```
Comment (OctoberChang) [inline pecos/core/ann/hnsw.hpp:1135]: should we also save other arguments passed into the train interface (e.g., `subspace_dimension`)?

File: pecos/core/ann/quantizer_impl/common.hpp [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+                    std::memcpy(xslice.data() + i * local_dimension, X_trn.val + index * dimension + m * local_dimension, local_dimension * sizeof(float));
+                }
+                pecos::drm_t Xsub;
+                Xsub.rows = sub_sample_points;
+                Xsub.cols = local_dimension;
+                Xsub.val = xslice.data();
+
+                // fit HLT or flat-Kmeans for each sub-space
+                std::vector<index_type> assignments(sub_sample_points);
+                pecos::clustering::Tree hlt(4);
```
Comment (OctoberChang) [inline pecos/core/ann/quantizer_impl/common.hpp]: avoid hard-coded, try `log2(num_of_local_centroids)`.

File: pecos/core/ann/hnsw.hpp [hunk 4]
```diff
@@ -870,5 +1009,343 @@ namespace ann {
     };
 
 
-} // end of namespace ann
-} // end of namespace pecos
+    template<typename dist_t>
+    struct HNSWProductQuantizer4Bits {
+        typedef FeatVecDenseL2Simd<float> feat_vec_t;
```
Comment (OctoberChang) [inline pecos/core/ann/hnsw.hpp]: we currently only support `L2`, need to add some warning.

File: pecos/core/ann/hnsw.hpp [hunk 5]
```diff
@@ -870,5 +1009,343 @@ namespace ann {
...
+            graph_l1.num_node = hnsw.graph_l1.num_node;
+            graph_l1.max_level = hnsw.graph_l1.max_level;
+            graph_l1.max_degree = hnsw.graph_l1.max_degree;
+            graph_l1.node_mem_size = hnsw.graph_l1.node_mem_size;
+            graph_l1.level_mem_size = hnsw.graph_l1.level_mem_size;
+            graph_l1.buffer.resize(hnsw.graph_l1.buffer.size());
+            memcpy(graph_l1.buffer.data(), hnsw.graph_l1.buffer.data(), hnsw.graph_l1.buffer.size() * sizeof(index_type));
+
+            graph_l0_pq4.build_quantizer(X_trn, subspace_dimension, sub_sample_points);
+            graph_l0_pq4.build_graph(hnsw.graph_l0);
```
Comment (OctoberChang) [inline pecos/core/ann/hnsw.hpp]: should we `delete hnsw` to reduce data from 3x to 2x?

File: pecos/core/ann/quantizer_impl/common.hpp (hunk 3 above)
Comment (Patrick-H-Chen) [reply pecos/core/ann/quantizer_impl/common.hpp]: Resolved in next commit.

File: pecos/core/ann/hnsw.hpp (hunk 2 above)
Comment (Patrick-H-Chen) [reply pecos/core/ann/hnsw.hpp:1135]: Resolved in next commit.

File: pecos/core/ann/hnsw.hpp (hunk 1 above)
Comment (Patrick-H-Chen) [reply pecos/core/ann/hnsw.hpp:246]: We will firstly restrict the implementation to FeatVecDenseL2Simd<float>, and extend it later when other metric is implemented.

File: pecos/core/ann/hnsw.hpp (hunk 4 above)
Comment (Patrick-H-Chen) [reply pecos/core/ann/hnsw.hpp]: Currently we don't have parameter to let user input distance metric. Maybe it's easier to write it in the README.md in the example folder ?

File: pecos/core/ann/quantizer_impl/x86.hpp [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+            approximate_neighbor_group_distance_default(neighbor_size, ds, neighbor_codes, lut_ptr, scale, bias);
+        }
+
+        __attribute__((__target__("default")))
+        inline void setup_lut(float* query, uint8_t* lut, float& scale, float& bias) const {
+            setup_lut_default(query, lut, scale, bias);
+        }
+
+        __attribute__((__target__("avx512f")))
+        inline void setup_lut(float* query, uint8_t* lut_ptr, float& scale, float& bias) const {
```
Comment (OctoberChang) [inline pecos/core/ann/quantizer_impl/x86.hpp:160]: consider renaming `lut_ptr` to `lut` (as in Line 154) or vice versa for naming consistency.

File: pecos/core/ann/quantizer_impl/x86.hpp [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+                }
+            }
+        }
+
+        __attribute__((__target__("default")))
+        void pad_parameters(index_type& max_degree, index_type& code_dimension) {
+            pad_parameters_default(max_degree, code_dimension);
+        }
+
+        __attribute__((__target__("avx512f"))) void pad_parameters(index_type& max_degree, index_type& code_dimension) {
```
Comment (OctoberChang) [inline pecos/core/ann/quantizer_impl/x86.hpp]: consider a newline before `void pad_parameters()`

File: pecos/core/ann/quantizer_impl/x86.hpp [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+            //  When using AVX512f, we have 16 centroids per local codebook, and each of it uses 8 bits to represent quantized
+            //  distance value. Thus,m we will have 128 bits to load 1 set of local codebooks. Thus, a loadu_si512 will load
```
Comment (OctoberChang) [inline pecos/core/ann/quantizer_impl/x86.hpp]: typo: `Thus,m` -> `Thus,`

File: pecos/core/ann/quantizer_impl/x86.hpp [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+                distance = _mm512_mul_ps(distance, _scale);
+                distance = _mm512_add_ps(distance, _bias);
+
+                _mm512_storeu_ps(d, distance);
+                d += 16;
+            }
+        }
+
+        __attribute__((__target__("default")))
+        inline void approximate_neighbor_group_distance(size_t neighbor_size, float* ds, const char* neighbor_codes, uint8_t* lut_ptr, float scale, float bias) const {
```
Comment (OctoberChang) [inline pecos/core/ann/quantizer_impl/x86.hpp:150]: naming consistency: `lut_ptr` or `lut`?

File: pecos/core/ann/quantizer_impl/x86.hpp [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+            // AVX512f will process 16 operations in parallel, and due to memory layout issues, to achieve fast memory load,
+            // we have to process all dimensions of 16 members first then move to next 16 neighbors. Therefore, we have to make
+            // sure the max_degree is a multiple of 16 so there will be no segmentation faults when reading codes from graph.
+            max_degree = max_degree % 16 == 0 ? max_degree : (max_degree / 16 + 1) * 16;
+        }
+
+        __attribute__((__target__("avx512f")))
+        inline void approximate_neighbor_group_distance(size_t neighbor_size, float* ds, const char* neighbor_codes, uint8_t* lut_ptr, float scale, float bias) const {
+            //  When using AVX512f, we have 16 centroids per local codebook, and each of it uses 8 bits to represent quantized
+            //  distance value. Thus,m we will have 128 bits to load 1 set of local codebooks. Thus, a loadu_si512 will load
```
Comment (OctoberChang) [inline pecos/core/ann/quantizer_impl/x86.hpp]: typo: `Thus,m` -> `Thus,`

File: pecos/core/ann/quantizer_impl/x86.hpp [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+            //  512 / 128 == 4 local codebooks at a time. Thus, number of loadu_si512 perfomred on a group will need to be
+            //  adjusted baed on if num_local_codebooks is divisible by 4.
+            index_type num_dimension_block = num_local_codebooks % 4 == 0 ? num_local_codebooks / 4 : num_local_codebooks / 4 + 1;
+
+            // Similarly, we have to parse every 16 neighbors at a time to maximally leverage avx512f.
+            // Thus we can also calculate the number of group iterations based on if neighbor_size id
+            // divisible by 16.
+            index_type num_groups = neighbor_size % 16 == 0 ? neighbor_size / 16 : neighbor_size / 16 + 1;
```
Comment (OctoberChang) [inline pecos/core/ann/quantizer_impl/x86.hpp:101]: This seems to be handled during Line 86 (the `pad_parameters()` function) ? If so, we can just do `num_groups=neighbor_size % 16`?

File: pecos/core/ann/quantizer_impl/common.hpp [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+                }
+            }
+        }
+
+        inline void train(const pecos::drm_t& X_trn, int M, size_t sub_sample_points=0, int seed=0, size_t max_iter=10, int threads=32) {
+            size_t dimension = X_trn.cols;
+            if (dimension % M != 0) {
+                throw std::runtime_error("Original dimension must be divided by subspace dimension");
+            }
+            num_local_codebooks = M;
```
Comment (OctoberChang) [inline pecos/core/ann/quantizer_impl/common.hpp]: consider just replace `M` by `num_local_codebooks`, and use that naming consistently in other place.

File: pecos/core/ann/quantizer_impl/x86.hpp [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+        void pad_parameters(index_type& max_degree, index_type& code_dimension) {
+            pad_parameters_default(max_degree, code_dimension);
+        }
+
+        __attribute__((__target__("avx512f"))) void pad_parameters(index_type& max_degree, index_type& code_dimension) {
+            //  When using AVX512f, we have 16 centroids per local codebook, and each of it uses 8 bits to represent quantized
+            //  distance value. Thus,m we will have 128 bits to load 1 set of local codebooks. Thus, a loadu_si512 will load
+            //  512 / 128 == 4 local codebooks at a time. Thus, will need to be adjust the code_dimension to make it divisible
+            //  by 4. If it's not, we have to pad 0 to extended dimensions.
+            code_dimension = code_dimension % 4 == 0 ? code_dimension : (code_dimension / 4 + 1) * 4;
```
Comment (OctoberChang) [inline pecos/core/ann/quantizer_impl/x86.hpp:82]: what's the relation between `code_dimension` versus `local_dimension` ?

File: pecos/core/ann/hnsw.hpp [hunk 14]
```diff
@@ -238,6 +243,143 @@ namespace ann {
...
+            if (subspace_dimension == 0) {
+                if (code_dimension >= 400) {
+                    code_dimension =  code_dimension % 2 == 0 ? code_dimension / 2 : code_dimension / 2 + 1;
+                }
+            } else {
+                code_dimension = code_dimension / subspace_dimension;
+            }
+            // currently, we don't support padding 0 on X_trn, so the cols of X_trn must be divisible by subspace_dimension.
+            // otherwise, we will throw error in quantizer.train().
+            quantizer.train(X_trn, code_dimension, sub_sample_points);
```
Comment (OctoberChang) [inline pecos/core/ann/hnsw.hpp:318]: in this case, `code_dimension` is actually `num_local_codebook` ?!

Comment (OctoberChang) [reply pecos/core/ann/hnsw.hpp:318]: consider naming variables for consistency in `quantizer.hpp`

File: pecos/core/ann/hnsw.hpp [hunk 15]
```diff
@@ -238,6 +243,143 @@ namespace ann {
         }
     };
 
+    struct GraphProductQuantizer4Bits : GraphBase {
+        typedef FeatVecDenseL2Simd<float> feat_vec_t;
+
+        ProductQuantizer4Bits quantizer;
+        index_type num_node;
+        index_type code_dimension;
```
Comment (OctoberChang) [inline pecos/core/ann/hnsw.hpp]: is this equivalent to `local_dimension` as defined `quantizer struct`?

File: pecos/core/ann/hnsw.hpp (hunk 14 above)
Comment (Patrick-H-Chen) [reply pecos/core/ann/hnsw.hpp:318]: code_dimension is from the perspective of HNSWPQ4Bits and num_local_codebook is from the perspective of Quantizer. So code_dimension could be padded to fit memory layout in GraphHNSWPQ4 but num_local_codebook will not be padded.

File: pecos/core/ann/quantizer_impl/x86.hpp (hunk 11 above)
Comment (Patrick-H-Chen) [reply pecos/core/ann/quantizer_impl/x86.hpp:101]: num_groups depends on neighbor_size, and Line 86 consider the maximal neighbors (max_degree). So it's possible max neighbors is 96, but for a node it only has say 18 neighbors, which is not divisible by 16. So we still need to check again here.

File: pecos/core/ann/quantizer_impl/x86.hpp (hunk 13 above)
Comment (Patrick-H-Chen) [reply pecos/core/ann/quantizer_impl/x86.hpp:82]: Same as below, code_dimension is from the perspective of HNSWPQ4Bits and local_dimension is from the perspective of quantizer. I will not pad  local_dimension but code_dimension could be padded if to fit memory layout in GraphPQ4Bits. In sum, code_dimension could be different from local_dimension.

File: pecos/core/ann/quantizer_impl/x86.hpp (hunk 6 above)
Comment (Patrick-H-Chen) [reply pecos/core/ann/quantizer_impl/x86.hpp:160]: lut refers to std::vector<float> and lut_ptr is lut.data(). If I don't use it as a function argument, anyway I need to initiate a pointer to traverse the lut.

File: pecos/core/ann/hnsw.hpp (hunk 14 above)
Comment (Patrick-H-Chen) [reply pecos/core/ann/hnsw.hpp:318]: After reading again the code, I think renaming code_dimension into num_local_codebook_pad is actually not a good idea. From Graph construction perspective, it's the number of dimensions a code will use, and there is no concept of "local codebook" from Graph's view. In addition, when calculating the memory space for neighboring code, code_dimension is a much more meaningful name than num_local_codebook_pad. So I think the name could be kept same.

File: pecos/core/ann/quantizer_impl/common.hpp (hunk 12 above)
Comment (Patrick-H-Chen) [reply pecos/core/ann/quantizer_impl/common.hpp]: Resolve in the next commit.

File: pecos/core/ann/quantizer_impl/x86.hpp (hunk 9 above)
Comment (Patrick-H-Chen) [reply pecos/core/ann/quantizer_impl/x86.hpp:150]: resolve in the next commit.

File: pecos/core/ann/hnsw.hpp (hunk 15 above)
Comment (Patrick-H-Chen) [reply pecos/core/ann/hnsw.hpp]: local_dimension is the number of dimensions of subspace, but code_dimension is the code length encoded in the graph for a single data point. code_dimension is the padded num_local_codebooks, and as explained above I think it's better not to rename.

File: pecos/core/ann/hnsw.hpp (hunk 14 above)
Comment (OctoberChang) [reply pecos/core/ann/hnsw.hpp:318]: I see. In such case, I would suggest you write some comments that describe the definitions of those variables under those structures (e.g., GraphL0PQ4bit, HNSWProdQuad4Bit, quantizer)

Comment (Patrick-H-Chen) [reply pecos/core/ann/hnsw.hpp:318]: Revise in the next commit.

**Review by OctoberChang — APPROVED** [review]

**Pull request merged**